// adb-ts/lib/util/keycode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyCode = void 0;
var KeyCode;
(function (KeyCode) {
    KeyCode[KeyCode["KEYCODE_UNKNOWN"] = 0] = "KEYCODE_UNKNOWN";
    /** Key code constant: Soft Left key.
     * Usually situated below the display on phones and used as a multi-function
     * feature key for selecting a software defined function shown on the bottom left
     * of the display. */
    KeyCode[KeyCode["KEYCODE_SOFT_LEFT"] = 1] = "KEYCODE_SOFT_LEFT";
    /** Key code constant: Soft Right key.
     * Usually situated below the display on phones and used as a multi-function
     * feature key for selecting a software defined function shown on the bottom right
     * of the display. */
    KeyCode[KeyCode["KEYCODE_SOFT_RIGHT"] = 2] = "KEYCODE_SOFT_RIGHT";
    /** Key code constant: Home key.
     * This key is handled by the framework and is never delivered to applications. */
    KeyCode[KeyCode["KEYCODE_HOME"] = 3] = "KEYCODE_HOME";
    /** Key code constant: Back key. */
    KeyCode[KeyCode["KEYCODE_BACK"] = 4] = "KEYCODE_BACK";
    /** Key code constant: Call key. */
    KeyCode[KeyCode["KEYCODE_CALL"] = 5] = "KEYCODE_CALL";
    /** Key code constant: End Call key. */
    KeyCode[KeyCode["KEYCODE_ENDCALL"] = 6] = "KEYCODE_ENDCALL";
    /** Key code constant: '0' key. */
    KeyCode[KeyCode["KEYCODE_0"] = 7] = "KEYCODE_0";
    /** Key code constant: '1' key. */
    KeyCode[KeyCode["KEYCODE_1"] = 8] = "KEYCODE_1";
    /** Key code constant: '2' key. */
    KeyCode[KeyCode["KEYCODE_2"] = 9] = "KEYCODE_2";
    /** Key code constant: '3' key. */
    KeyCode[KeyCode["KEYCODE_3"] = 10] = "KEYCODE_3";
    /** Key code constant: '4' key. */
    KeyCode[KeyCode["KEYCODE_4"] = 11] = "KEYCODE_4";
    /** Key code constant: '5' key. */
    KeyCode[KeyCode["KEYCODE_5"] = 12] = "KEYCODE_5";
    /** Key code constant: '6' key. */
    KeyCode[KeyCode["KEYCODE_6"] = 13] = "KEYCODE_6";
    /** Key code constant: '7' key. */
    KeyCode[KeyCode["KEYCODE_7"] = 14] = "KEYCODE_7";
    /** Key code constant: '8' key. */
    KeyCode[KeyCode["KEYCODE_8"] = 15] = "KEYCODE_8";
    /** Key code constant: '9' key. */
    KeyCode[KeyCode["KEYCODE_9"] = 16] = "KEYCODE_9";
    /** Key code constant: '*' key. */
    KeyCode[KeyCode["KEYCODE_STAR"] = 17] = "KEYCODE_STAR";
    /** Key code constant: '#' key. */
    KeyCode[KeyCode["KEYCODE_POUND"] = 18] = "KEYCODE_POUND";
    /** Key code constant: Directional Pad Up key.
     * May also be synthesized from trackball motions. */
    KeyCode[KeyCode["KEYCODE_DPAD_UP"] = 19] = "KEYCODE_DPAD_UP";
    /** Key code constant: Directional Pad Down key.
     * May also be synthesized from trackball motions. */
    KeyCode[KeyCode["KEYCODE_DPAD_DOWN"] = 20] = "KEYCODE_DPAD_DOWN";
    /** Key code constant: Directional Pad Left key.
     * May also be synthesized from trackball motions. */
    KeyCode[KeyCode["KEYCODE_DPAD_LEFT"] = 21] = "KEYCODE_DPAD_LEFT";
    /** Key code constant: Directional Pad Right key.
     * May also be synthesized from trackball motions. */
    KeyCode[KeyCode["KEYCODE_DPAD_RIGHT"] = 22] = "KEYCODE_DPAD_RIGHT";
    /** Key code constant: Directional Pad Center key.
     * May also be synthesized from trackball motions. */
    KeyCode[KeyCode["KEYCODE_DPAD_CENTER"] = 23] = "KEYCODE_DPAD_CENTER";
    /** Key code constant: Volume Up key.
     * Adjusts the speaker volume up. */
    KeyCode[KeyCode["KEYCODE_VOLUME_UP"] = 24] = "KEYCODE_VOLUME_UP";
    /** Key code constant: Volume Down key.
     * Adjusts the speaker volume down. */
    KeyCode[KeyCode["KEYCODE_VOLUME_DOWN"] = 25] = "KEYCODE_VOLUME_DOWN";
    /** Key code constant: Power key. */
    KeyCode[KeyCode["KEYCODE_POWER"] = 26] = "KEYCODE_POWER";
    /** Key code constant: Camera key.
     * Used to launch a camera application or take pictures. */
    KeyCode[KeyCode["KEYCODE_CAMERA"] = 27] = "KEYCODE_CAMERA";
    /** Key code constant: Clear key. */
    KeyCode[KeyCode["KEYCODE_CLEAR"] = 28] = "KEYCODE_CLEAR";
    /** Key code constant: 'A' key. */
    KeyCode[KeyCode["KEYCODE_A"] = 29] = "KEYCODE_A";
    /** Key code constant: 'B' key. */
    KeyCode[KeyCode["KEYCODE_B"] = 30] = "KEYCODE_B";
    /** Key code constant: 'C' key. */
    KeyCode[KeyCode["KEYCODE_C"] = 31] = "KEYCODE_C";
    /** Key code constant: 'D' key. */
    KeyCode[KeyCode["KEYCODE_D"] = 32] = "KEYCODE_D";
    /** Key code constant: 'E' key. */
    KeyCode[KeyCode["KEYCODE_E"] = 33] = "KEYCODE_E";
    /** Key code constant: 'F' key. */
    KeyCode[KeyCode["KEYCODE_F"] = 34] = "KEYCODE_F";
    /** Key code constant: 'G' key. */
    KeyCode[KeyCode["KEYCODE_G"] = 35] = "KEYCODE_G";
    /** Key code constant: 'H' key. */
    KeyCode[KeyCode["KEYCODE_H"] = 36] = "KEYCODE_H";
    /** Key code constant: 'I' key. */
    KeyCode[KeyCode["KEYCODE_I"] = 37] = "KEYCODE_I";
    /** Key code constant: 'J' key. */
    KeyCode[KeyCode["KEYCODE_J"] = 38] = "KEYCODE_J";
    /** Key code constant: 'K' key. */
    KeyCode[KeyCode["KEYCODE_K"] = 39] = "KEYCODE_K";
    /** Key code constant: 'L' key. */
    KeyCode[KeyCode["KEYCODE_L"] = 40] = "KEYCODE_L";
    /** Key code constant: 'M' key. */
    KeyCode[KeyCode["KEYCODE_M"] = 41] = "KEYCODE_M";
    /** Key code constant: 'N' key. */
    KeyCode[KeyCode["KEYCODE_N"] = 42] = "KEYCODE_N";
    /** Key code constant: 'O' key. */
    KeyCode[KeyCode["KEYCODE_O"] = 43] = "KEYCODE_O";
    /** Key code constant: 'P' key. */
    KeyCode[KeyCode["KEYCODE_P"] = 44] = "KEYCODE_P";
    /** Key code constant: 'Q' key. */
    KeyCode[KeyCode["KEYCODE_Q"] = 45] = "KEYCODE_Q";
    /** Key code constant: 'R' key. */
    KeyCode[KeyCode["KEYCODE_R"] = 46] = "KEYCODE_R";
    /** Key code constant: 'S' key. */
    KeyCode[KeyCode["KEYCODE_S"] = 47] = "KEYCODE_S";
    /** Key code constant: 'T' key. */
    KeyCode[KeyCode["KEYCODE_T"] = 48] = "KEYCODE_T";
    /** Key code constant: 'U' key. */
    KeyCode[KeyCode["KEYCODE_U"] = 49] = "KEYCODE_U";
    /** Key code constant: 'V' key. */
    KeyCode[KeyCode["KEYCODE_V"] = 50] = "KEYCODE_V";
    /** Key code constant: 'W' key. */
    KeyCode[KeyCode["KEYCODE_W"] = 51] = "KEYCODE_W";
    /** Key code constant: 'X' key. */
    KeyCode[KeyCode["KEYCODE_X"] = 52] = "KEYCODE_X";
    /** Key code constant: 'Y' key. */
    KeyCode[KeyCode["KEYCODE_Y"] = 53] = "KEYCODE_Y";
    /** Key code constant: 'Z' key. */
    KeyCode[KeyCode["KEYCODE_Z"] = 54] = "KEYCODE_Z";
    /** Key code constant: ',' key. */
    KeyCode[KeyCode["KEYCODE_COMMA"] = 55] = "KEYCODE_COMMA";
    /** Key code constant: '.' key. */
    KeyCode[KeyCode["KEYCODE_PERIOD"] = 56] = "KEYCODE_PERIOD";
    /** Key code constant: Left Alt modifier key. */
    KeyCode[KeyCode["KEYCODE_ALT_LEFT"] = 57] = "KEYCODE_ALT_LEFT";
    /** Key code constant: Right Alt modifier key. */
    KeyCode[KeyCode["KEYCODE_ALT_RIGHT"] = 58] = "KEYCODE_ALT_RIGHT";
    /** Key code constant: Left Shift modifier key. */
    KeyCode[KeyCode["KEYCODE_SHIFT_LEFT"] = 59] = "KEYCODE_SHIFT_LEFT";
    /** Key code constant: Right Shift modifier key. */
    KeyCode[KeyCode["KEYCODE_SHIFT_RIGHT"] = 60] = "KEYCODE_SHIFT_RIGHT";
    /** Key code constant: Tab key. */
    KeyCode[KeyCode["KEYCODE_TAB"] = 61] = "KEYCODE_TAB";
    /** Key code constant: Space key. */
    KeyCode[KeyCode["KEYCODE_SPACE"] = 62] = "KEYCODE_SPACE";
    /** Key code constant: Symbol modifier key.
     * Used to enter alternate symbols. */
    KeyCode[KeyCode["KEYCODE_SYM"] = 63] = "KEYCODE_SYM";
    /** Key code constant: Explorer special function key.
     * Used to launch a browser application. */
    KeyCode[KeyCode["KEYCODE_EXPLORER"] = 64] = "KEYCODE_EXPLORER";
    /** Key code constant: Envelope special function key.
     * Used to launch a mail application. */
    KeyCode[KeyCode["KEYCODE_ENVELOPE"] = 65] = "KEYCODE_ENVELOPE";
    /** Key code constant: Enter key. */
    KeyCode[KeyCode["KEYCODE_ENTER"] = 66] = "KEYCODE_ENTER";
    /** Key code constant: Backspace key.
     * Deletes characters before the insertion po , unlike `KEYCODE_FORWARD_DEL`. */
    KeyCode[KeyCode["KEYCODE_DEL"] = 67] = "KEYCODE_DEL";
    /** Key code constant: '`' (backtick) key. */
    KeyCode[KeyCode["KEYCODE_GRAVE"] = 68] = "KEYCODE_GRAVE";
    /** Key code constant: '-'. */
    KeyCode[KeyCode["KEYCODE_MINUS"] = 69] = "KEYCODE_MINUS";
    /** Key code constant: '=' key. */
    KeyCode[KeyCode["KEYCODE_EQUALS"] = 70] = "KEYCODE_EQUALS";
    /** Key code constant: '[' key. */
    KeyCode[KeyCode["KEYCODE_LEFT_BRACKET"] = 71] = "KEYCODE_LEFT_BRACKET";
    /** Key code constant: ']' key. */
    KeyCode[KeyCode["KEYCODE_RIGHT_BRACKET"] = 72] = "KEYCODE_RIGHT_BRACKET";
    /** Key code constant: '\' key. */
    KeyCode[KeyCode["KEYCODE_BACKSLASH"] = 73] = "KEYCODE_BACKSLASH";
    /** Key code constant: ' ,' key. */
    KeyCode[KeyCode["KEYCODE_SEMICOLON"] = 74] = "KEYCODE_SEMICOLON";
    /** Key code constant: ''' (apostrophe) key. */
    KeyCode[KeyCode["KEYCODE_APOSTROPHE"] = 75] = "KEYCODE_APOSTROPHE";
    /** Key code constant: '/' key. */
    KeyCode[KeyCode["KEYCODE_SLASH"] = 76] = "KEYCODE_SLASH";
    /** Key code constant: '@' key. */
    KeyCode[KeyCode["KEYCODE_AT"] = 77] = "KEYCODE_AT";
    /** Key code constant: Number modifier key.
     * Used to enter numeric symbols.
     * This key is not Num Lock , it is more like `KEYCODE_ALT_LEFT` and is
     *  interpreted as an ALT key by `android.text.method.MetaKeyKeyListener`. */
    KeyCode[KeyCode["KEYCODE_NUM"] = 78] = "KEYCODE_NUM";
    /** Key code constant: Headset Hook key.
     * Used to hang up calls and stop media. */
    KeyCode[KeyCode["KEYCODE_HEADSETHOOK"] = 79] = "KEYCODE_HEADSETHOOK";
    /** Key code constant: Camera Focus key.
     * Used to focus the camera. */
    KeyCode[KeyCode["KEYCODE_FOCUS"] = 80] = "KEYCODE_FOCUS";
    /** Key code constant: '+' key. */
    KeyCode[KeyCode["KEYCODE_PLUS"] = 81] = "KEYCODE_PLUS";
    /** Key code constant: Menu key. */
    KeyCode[KeyCode["KEYCODE_MENU"] = 82] = "KEYCODE_MENU";
    /** Key code constant: Notification key. */
    KeyCode[KeyCode["KEYCODE_NOTIFICATION"] = 83] = "KEYCODE_NOTIFICATION";
    /** Key code constant: Search key. */
    KeyCode[KeyCode["KEYCODE_SEARCH"] = 84] = "KEYCODE_SEARCH";
    /** Key code constant: Play/Pause media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_PLAY_PAUSE"] = 85] = "KEYCODE_MEDIA_PLAY_PAUSE";
    /** Key code constant: Stop media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_STOP"] = 86] = "KEYCODE_MEDIA_STOP";
    /** Key code constant: Play Next media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_NEXT"] = 87] = "KEYCODE_MEDIA_NEXT";
    /** Key code constant: Play Previous media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_PREVIOUS"] = 88] = "KEYCODE_MEDIA_PREVIOUS";
    /** Key code constant: Rewind media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_REWIND"] = 89] = "KEYCODE_MEDIA_REWIND";
    /** Key code constant: Fast Forward media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_FAST_FORWARD"] = 90] = "KEYCODE_MEDIA_FAST_FORWARD";
    /** Key code constant: Mute key.
     * Mutes the microphone, unlike `KEYCODE_VOLUME_MUTE`. */
    KeyCode[KeyCode["KEYCODE_MUTE"] = 91] = "KEYCODE_MUTE";
    /** Key code constant: Page Up key. */
    KeyCode[KeyCode["KEYCODE_PAGE_UP"] = 92] = "KEYCODE_PAGE_UP";
    /** Key code constant: Page Down key. */
    KeyCode[KeyCode["KEYCODE_PAGE_DOWN"] = 93] = "KEYCODE_PAGE_DOWN";
    /** Key code constant: Picture Symbols modifier key.
     * Used to switch symbol sets (Emoji, Kao-moji). */
    KeyCode[KeyCode["KEYCODE_PICTSYMBOLS"] = 94] = "KEYCODE_PICTSYMBOLS";
    /** Key code constant: Switch Charset modifier key.
     * Used to switch character sets (Kanji, Katakana). */
    KeyCode[KeyCode["KEYCODE_SWITCH_CHARSET"] = 95] = "KEYCODE_SWITCH_CHARSET";
    /** Key code constant: A Button key.
     * On a game controller, the A button should be either the button labeled A
     * or the first button on the bottom row of controller buttons. */
    KeyCode[KeyCode["KEYCODE_BUTTON_A"] = 96] = "KEYCODE_BUTTON_A";
    /** Key code constant: B Button key.
     * On a game controller, the B button should be either the button labeled B
     * or the second button on the bottom row of controller buttons. */
    KeyCode[KeyCode["KEYCODE_BUTTON_B"] = 97] = "KEYCODE_BUTTON_B";
    /** Key code constant: C Button key.
     * On a game controller, the C button should be either the button labeled C
     * or the third button on the bottom row of controller buttons. */
    KeyCode[KeyCode["KEYCODE_BUTTON_C"] = 98] = "KEYCODE_BUTTON_C";
    /** Key code constant: X Button key.
     * On a game controller, the X button should be either the button labeled X
     * or the first button on the upper row of controller buttons. */
    KeyCode[KeyCode["KEYCODE_BUTTON_X"] = 99] = "KEYCODE_BUTTON_X";
    /** Key code constant: Y Button key.
     * On a game controller, the Y button should be either the button labeled Y
     * or the second button on the upper row of controller buttons. */
    KeyCode[KeyCode["KEYCODE_BUTTON_Y"] = 100] = "KEYCODE_BUTTON_Y";
    /** Key code constant: Z Button key.
     * On a game controller, the Z button should be either the button labeled Z
     * or the third button on the upper row of controller buttons. */
    KeyCode[KeyCode["KEYCODE_BUTTON_Z"] = 101] = "KEYCODE_BUTTON_Z";
    /** Key code constant: L1 Button key.
     * On a game controller, the L1 button should be either the button labeled L1 (or L)
     * or the top left trigger button. */
    KeyCode[KeyCode["KEYCODE_BUTTON_L1"] = 102] = "KEYCODE_BUTTON_L1";
    /** Key code constant: R1 Button key.
     * On a game controller, the R1 button should be either the button labeled R1 (or R)
     * or the top right trigger button. */
    KeyCode[KeyCode["KEYCODE_BUTTON_R1"] = 103] = "KEYCODE_BUTTON_R1";
    /** Key code constant: L2 Button key.
     * On a game controller, the L2 button should be either the button labeled L2
     * or the bottom left trigger button. */
    KeyCode[KeyCode["KEYCODE_BUTTON_L2"] = 104] = "KEYCODE_BUTTON_L2";
    /** Key code constant: R2 Button key.
     * On a game controller, the R2 button should be either the button labeled R2
     * or the bottom right trigger button. */
    KeyCode[KeyCode["KEYCODE_BUTTON_R2"] = 105] = "KEYCODE_BUTTON_R2";
    /** Key code constant: Left Thumb Button key.
     * On a game controller, the left thumb button indicates that the left (or only)
     * joystick is pressed. */
    KeyCode[KeyCode["KEYCODE_BUTTON_THUMBL"] = 106] = "KEYCODE_BUTTON_THUMBL";
    /** Key code constant: Right Thumb Button key.
     * On a game controller, the right thumb button indicates that the right
     * joystick is pressed. */
    KeyCode[KeyCode["KEYCODE_BUTTON_THUMBR"] = 107] = "KEYCODE_BUTTON_THUMBR";
    /** Key code constant: Start Button key.
     * On a game controller, the button labeled Start. */
    KeyCode[KeyCode["KEYCODE_BUTTON_START"] = 108] = "KEYCODE_BUTTON_START";
    /** Key code constant: Select Button key.
     * On a game controller, the button labeled Select. */
    KeyCode[KeyCode["KEYCODE_BUTTON_SELECT"] = 109] = "KEYCODE_BUTTON_SELECT";
    /** Key code constant: Mode Button key.
     * On a game controller, the button labeled Mode. */
    KeyCode[KeyCode["KEYCODE_BUTTON_MODE"] = 110] = "KEYCODE_BUTTON_MODE";
    /** Key code constant: Escape key. */
    KeyCode[KeyCode["KEYCODE_ESCAPE"] = 111] = "KEYCODE_ESCAPE";
    /** Key code constant: Forward Delete key.
     * Deletes characters ahead of the insertion po , unlike `KEYCODE_DEL`. */
    KeyCode[KeyCode["KEYCODE_FORWARD_DEL"] = 112] = "KEYCODE_FORWARD_DEL";
    /** Key code constant: Left Control modifier key. */
    KeyCode[KeyCode["KEYCODE_CTRL_LEFT"] = 113] = "KEYCODE_CTRL_LEFT";
    /** Key code constant: Right Control modifier key. */
    KeyCode[KeyCode["KEYCODE_CTRL_RIGHT"] = 114] = "KEYCODE_CTRL_RIGHT";
    /** Key code constant: Caps Lock key. */
    KeyCode[KeyCode["KEYCODE_CAPS_LOCK"] = 115] = "KEYCODE_CAPS_LOCK";
    /** Key code constant: Scroll Lock key. */
    KeyCode[KeyCode["KEYCODE_SCROLL_LOCK"] = 116] = "KEYCODE_SCROLL_LOCK";
    /** Key code constant: Left Meta modifier key. */
    KeyCode[KeyCode["KEYCODE_META_LEFT"] = 117] = "KEYCODE_META_LEFT";
    /** Key code constant: Right Meta modifier key. */
    KeyCode[KeyCode["KEYCODE_META_RIGHT"] = 118] = "KEYCODE_META_RIGHT";
    /** Key code constant: Function modifier key. */
    KeyCode[KeyCode["KEYCODE_FUNCTION"] = 119] = "KEYCODE_FUNCTION";
    /** Key code constant: System Request / Pr  Screen key. */
    KeyCode[KeyCode["KEYCODE_SYSRQ"] = 120] = "KEYCODE_SYSRQ";
    /** Key code constant: Break / Pause key. */
    KeyCode[KeyCode["KEYCODE_BREAK"] = 121] = "KEYCODE_BREAK";
    /** Key code constant: Home Movement key.
     * Used for scrolling or moving the cursor around to the start of a line
     * or to the top of a list. */
    KeyCode[KeyCode["KEYCODE_MOVE_HOME"] = 122] = "KEYCODE_MOVE_HOME";
    /** Key code constant: End Movement key.
     * Used for scrolling or moving the cursor around to the end of a line
     * or to the bottom of a list. */
    KeyCode[KeyCode["KEYCODE_MOVE_END"] = 123] = "KEYCODE_MOVE_END";
    /** Key code constant: Insert key.
     * Toggles insert / overwrite edit mode. */
    KeyCode[KeyCode["KEYCODE_INSERT"] = 124] = "KEYCODE_INSERT";
    /** Key code constant: Forward key.
     * Navigates forward in the history stack.  Complement of `KEYCODE_BACK`. */
    KeyCode[KeyCode["KEYCODE_FORWARD"] = 125] = "KEYCODE_FORWARD";
    /** Key code constant: Play media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_PLAY"] = 126] = "KEYCODE_MEDIA_PLAY";
    /** Key code constant: Pause media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_PAUSE"] = 127] = "KEYCODE_MEDIA_PAUSE";
    /** Key code constant: Close media key.
     * May be used to close a CD tray, for example. */
    KeyCode[KeyCode["KEYCODE_MEDIA_CLOSE"] = 128] = "KEYCODE_MEDIA_CLOSE";
    /** Key code constant: Eject media key.
     * May be used to eject a CD tray, for example. */
    KeyCode[KeyCode["KEYCODE_MEDIA_EJECT"] = 129] = "KEYCODE_MEDIA_EJECT";
    /** Key code constant: Record media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_RECORD"] = 130] = "KEYCODE_MEDIA_RECORD";
    /** Key code constant: F1 key. */
    KeyCode[KeyCode["KEYCODE_F1"] = 131] = "KEYCODE_F1";
    /** Key code constant: F2 key. */
    KeyCode[KeyCode["KEYCODE_F2"] = 132] = "KEYCODE_F2";
    /** Key code constant: F3 key. */
    KeyCode[KeyCode["KEYCODE_F3"] = 133] = "KEYCODE_F3";
    /** Key code constant: F4 key. */
    KeyCode[KeyCode["KEYCODE_F4"] = 134] = "KEYCODE_F4";
    /** Key code constant: F5 key. */
    KeyCode[KeyCode["KEYCODE_F5"] = 135] = "KEYCODE_F5";
    /** Key code constant: F6 key. */
    KeyCode[KeyCode["KEYCODE_F6"] = 136] = "KEYCODE_F6";
    /** Key code constant: F7 key. */
    KeyCode[KeyCode["KEYCODE_F7"] = 137] = "KEYCODE_F7";
    /** Key code constant: F8 key. */
    KeyCode[KeyCode["KEYCODE_F8"] = 138] = "KEYCODE_F8";
    /** Key code constant: F9 key. */
    KeyCode[KeyCode["KEYCODE_F9"] = 139] = "KEYCODE_F9";
    /** Key code constant: F10 key. */
    KeyCode[KeyCode["KEYCODE_F10"] = 140] = "KEYCODE_F10";
    /** Key code constant: F11 key. */
    KeyCode[KeyCode["KEYCODE_F11"] = 141] = "KEYCODE_F11";
    /** Key code constant: F12 key. */
    KeyCode[KeyCode["KEYCODE_F12"] = 142] = "KEYCODE_F12";
    /** Key code constant: Num Lock key.
     * This is the Num Lock key , it is different from `KEYCODE_NUM`.
     * This key alters the behavior of other keys on the numeric keypad. */
    KeyCode[KeyCode["KEYCODE_NUM_LOCK"] = 143] = "KEYCODE_NUM_LOCK";
    /** Key code constant: Numeric keypad '0' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_0"] = 144] = "KEYCODE_NUMPAD_0";
    /** Key code constant: Numeric keypad '1' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_1"] = 145] = "KEYCODE_NUMPAD_1";
    /** Key code constant: Numeric keypad '2' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_2"] = 146] = "KEYCODE_NUMPAD_2";
    /** Key code constant: Numeric keypad '3' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_3"] = 147] = "KEYCODE_NUMPAD_3";
    /** Key code constant: Numeric keypad '4' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_4"] = 148] = "KEYCODE_NUMPAD_4";
    /** Key code constant: Numeric keypad '5' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_5"] = 149] = "KEYCODE_NUMPAD_5";
    /** Key code constant: Numeric keypad '6' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_6"] = 150] = "KEYCODE_NUMPAD_6";
    /** Key code constant: Numeric keypad '7' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_7"] = 151] = "KEYCODE_NUMPAD_7";
    /** Key code constant: Numeric keypad '8' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_8"] = 152] = "KEYCODE_NUMPAD_8";
    /** Key code constant: Numeric keypad '9' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_9"] = 153] = "KEYCODE_NUMPAD_9";
    /** Key code constant: Numeric keypad '/' key (for division). */
    KeyCode[KeyCode["KEYCODE_NUMPAD_DIVIDE"] = 154] = "KEYCODE_NUMPAD_DIVIDE";
    /** Key code constant: Numeric keypad '*' key (for multiplication). */
    KeyCode[KeyCode["KEYCODE_NUMPAD_MULTIPLY"] = 155] = "KEYCODE_NUMPAD_MULTIPLY";
    /** Key code constant: Numeric keypad '-' key (for subtraction). */
    KeyCode[KeyCode["KEYCODE_NUMPAD_SUBTRACT"] = 156] = "KEYCODE_NUMPAD_SUBTRACT";
    /** Key code constant: Numeric keypad '+' key (for addition). */
    KeyCode[KeyCode["KEYCODE_NUMPAD_ADD"] = 157] = "KEYCODE_NUMPAD_ADD";
    /** Key code constant: Numeric keypad '.' key (for decimals or digit grouping). */
    KeyCode[KeyCode["KEYCODE_NUMPAD_DOT"] = 158] = "KEYCODE_NUMPAD_DOT";
    /** Key code constant: Numeric keypad ',' key (for decimals or digit grouping). */
    KeyCode[KeyCode["KEYCODE_NUMPAD_COMMA"] = 159] = "KEYCODE_NUMPAD_COMMA";
    /** Key code constant: Numeric keypad Enter key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_ENTER"] = 160] = "KEYCODE_NUMPAD_ENTER";
    /** Key code constant: Numeric keypad '=' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_EQUALS"] = 161] = "KEYCODE_NUMPAD_EQUALS";
    /** Key code constant: Numeric keypad '(' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_LEFT_PAREN"] = 162] = "KEYCODE_NUMPAD_LEFT_PAREN";
    /** Key code constant: Numeric keypad ')' key. */
    KeyCode[KeyCode["KEYCODE_NUMPAD_RIGHT_PAREN"] = 163] = "KEYCODE_NUMPAD_RIGHT_PAREN";
    /** Key code constant: Volume Mute key.
     * Mutes the speaker, unlike `KEYCODE_MUTE`.
     * This key should normally be implemented as a toggle such that the first press
     * mutes the speaker and the second press restores the original volume. */
    KeyCode[KeyCode["KEYCODE_VOLUME_MUTE"] = 164] = "KEYCODE_VOLUME_MUTE";
    /** Key code constant: Info key.
     * Common on TV remotes to show additional information related to what is
     * currently being viewed. */
    KeyCode[KeyCode["KEYCODE_INFO"] = 165] = "KEYCODE_INFO";
    /** Key code constant: Channel up key.
     * On TV remotes, increments the television channel. */
    KeyCode[KeyCode["KEYCODE_CHANNEL_UP"] = 166] = "KEYCODE_CHANNEL_UP";
    /** Key code constant: Channel down key.
     * On TV remotes, decrements the television channel. */
    KeyCode[KeyCode["KEYCODE_CHANNEL_DOWN"] = 167] = "KEYCODE_CHANNEL_DOWN";
    /** Key code constant: Zoom in key. */
    KeyCode[KeyCode["KEYCODE_ZOOM_IN"] = 168] = "KEYCODE_ZOOM_IN";
    /** Key code constant: Zoom out key. */
    KeyCode[KeyCode["KEYCODE_ZOOM_OUT"] = 169] = "KEYCODE_ZOOM_OUT";
    /** Key code constant: TV key.
     * On TV remotes, switches to viewing live TV. */
    KeyCode[KeyCode["KEYCODE_TV"] = 170] = "KEYCODE_TV";
    /** Key code constant: Window key.
     * On TV remotes, toggles picture-in-picture mode or other windowing functions.
     * On Android Wear devices, triggers a display offset. */
    KeyCode[KeyCode["KEYCODE_WINDOW"] = 171] = "KEYCODE_WINDOW";
    /** Key code constant: Guide key.
     * On TV remotes, shows a programming guide. */
    KeyCode[KeyCode["KEYCODE_GUIDE"] = 172] = "KEYCODE_GUIDE";
    /** Key code constant: DVR key.
     * On some TV remotes, switches to a DVR mode for recorded shows. */
    KeyCode[KeyCode["KEYCODE_DVR"] = 173] = "KEYCODE_DVR";
    /** Key code constant: Bookmark key.
     * On some TV remotes, bookmarks content or web pages. */
    KeyCode[KeyCode["KEYCODE_BOOKMARK"] = 174] = "KEYCODE_BOOKMARK";
    /** Key code constant: Toggle captions key.
     * Switches the mode for closed-captioning text, for example during television shows. */
    KeyCode[KeyCode["KEYCODE_CAPTIONS"] = 175] = "KEYCODE_CAPTIONS";
    /** Key code constant: Settings key.
     * Starts the system settings activity. */
    KeyCode[KeyCode["KEYCODE_SETTINGS"] = 176] = "KEYCODE_SETTINGS";
    /** Key code constant: TV power key.
     * On TV remotes, toggles the power on a television screen. */
    KeyCode[KeyCode["KEYCODE_TV_POWER"] = 177] = "KEYCODE_TV_POWER";
    /** Key code constant: TV input key.
     * On TV remotes, switches the input on a television screen. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT"] = 178] = "KEYCODE_TV_INPUT";
    /** Key code constant: Set-top-box power key.
     * On TV remotes, toggles the power on an external Set-top-box. */
    KeyCode[KeyCode["KEYCODE_STB_POWER"] = 179] = "KEYCODE_STB_POWER";
    /** Key code constant: Set-top-box input key.
     * On TV remotes, switches the input mode on an external Set-top-box. */
    KeyCode[KeyCode["KEYCODE_STB_INPUT"] = 180] = "KEYCODE_STB_INPUT";
    /** Key code constant: A/V Receiver power key.
     * On TV remotes, toggles the power on an external A/V Receiver. */
    KeyCode[KeyCode["KEYCODE_AVR_POWER"] = 181] = "KEYCODE_AVR_POWER";
    /** Key code constant: A/V Receiver input key.
     * On TV remotes, switches the input mode on an external A/V Receiver. */
    KeyCode[KeyCode["KEYCODE_AVR_INPUT"] = 182] = "KEYCODE_AVR_INPUT";
    /** Key code constant: Red "programmable" key.
     * On TV remotes, acts as a contextual/programmable key. */
    KeyCode[KeyCode["KEYCODE_PROG_RED"] = 183] = "KEYCODE_PROG_RED";
    /** Key code constant: Green "programmable" key.
     * On TV remotes, actsas a contextual/programmable key. */
    KeyCode[KeyCode["KEYCODE_PROG_GREEN"] = 184] = "KEYCODE_PROG_GREEN";
    /** Key code constant: Yellow "programmable" key.
     * On TV remotes, acts as a contextual/programmable key. */
    KeyCode[KeyCode["KEYCODE_PROG_YELLOW"] = 185] = "KEYCODE_PROG_YELLOW";
    /** Key code constant: Blue "programmable" key.
     * On TV remotes, acts as a contextual/programmable key. */
    KeyCode[KeyCode["KEYCODE_PROG_BLUE"] = 186] = "KEYCODE_PROG_BLUE";
    /** Key code constant: App switch key.
     * Should bring up the application switcher dialog. */
    KeyCode[KeyCode["KEYCODE_APP_SWITCH"] = 187] = "KEYCODE_APP_SWITCH";
    /** Key code constant: Generic Game Pad Button #1.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_1"] = 188] = "KEYCODE_BUTTON_1";
    /** Key code constant: Generic Game Pad Button #2.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_2"] = 189] = "KEYCODE_BUTTON_2";
    /** Key code constant: Generic Game Pad Button #3.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_3"] = 190] = "KEYCODE_BUTTON_3";
    /** Key code constant: Generic Game Pad Button #4.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_4"] = 191] = "KEYCODE_BUTTON_4";
    /** Key code constant: Generic Game Pad Button #5.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_5"] = 192] = "KEYCODE_BUTTON_5";
    /** Key code constant: Generic Game Pad Button #6.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_6"] = 193] = "KEYCODE_BUTTON_6";
    /** Key code constant: Generic Game Pad Button #7.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_7"] = 194] = "KEYCODE_BUTTON_7";
    /** Key code constant: Generic Game Pad Button #8.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_8"] = 195] = "KEYCODE_BUTTON_8";
    /** Key code constant: Generic Game Pad Button #9.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_9"] = 196] = "KEYCODE_BUTTON_9";
    /** Key code constant: Generic Game Pad Button #10.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_10"] = 197] = "KEYCODE_BUTTON_10";
    /** Key code constant: Generic Game Pad Button #11.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_11"] = 198] = "KEYCODE_BUTTON_11";
    /** Key code constant: Generic Game Pad Button #12.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_12"] = 199] = "KEYCODE_BUTTON_12";
    /** Key code constant: Generic Game Pad Button #13.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_13"] = 200] = "KEYCODE_BUTTON_13";
    /** Key code constant: Generic Game Pad Button #14.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_14"] = 201] = "KEYCODE_BUTTON_14";
    /** Key code constant: Generic Game Pad Button #15.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_15"] = 202] = "KEYCODE_BUTTON_15";
    /** Key code constant: Generic Game Pad Button #16.*/
    KeyCode[KeyCode["KEYCODE_BUTTON_16"] = 203] = "KEYCODE_BUTTON_16";
    /** Key code constant: Language Switch key.
     * Toggles the current input language such as switching between English and Japanese on
     * a QWERTY keyboard.  On some devices, the same function may be performed by
     * pressing Shift+Spacebar. */
    KeyCode[KeyCode["KEYCODE_LANGUAGE_SWITCH"] = 204] = "KEYCODE_LANGUAGE_SWITCH";
    /** Key code constant: Manner Mode key.
     * Toggles silent or vibrate mode on and off to make the device behave more politely
     * in certain settings such as on a crowded train.  On some devices, the key may only
     * operate when long-pressed. */
    KeyCode[KeyCode["KEYCODE_MANNER_MODE"] = 205] = "KEYCODE_MANNER_MODE";
    /** Key code constant: 3D Mode key.
     * Toggles the display between 2D and 3D mode. */
    KeyCode[KeyCode["KEYCODE_3D_MODE"] = 206] = "KEYCODE_3D_MODE";
    /** Key code constant: Contacts special function key.
     * Used to launch an address book application. */
    KeyCode[KeyCode["KEYCODE_CONTACTS"] = 207] = "KEYCODE_CONTACTS";
    /** Key code constant: Calendar special function key.
     * Used to launch a calendar application. */
    KeyCode[KeyCode["KEYCODE_CALENDAR"] = 208] = "KEYCODE_CALENDAR";
    /** Key code constant: Music special function key.
     * Used to launch a music player application. */
    KeyCode[KeyCode["KEYCODE_MUSIC"] = 209] = "KEYCODE_MUSIC";
    /** Key code constant: Calculator special function key.
     * Used to launch a calculator application. */
    KeyCode[KeyCode["KEYCODE_CALCULATOR"] = 210] = "KEYCODE_CALCULATOR";
    /** Key code constant: Japanese full-width / half-width key. */
    KeyCode[KeyCode["KEYCODE_ZENKAKU_HANKAKU"] = 211] = "KEYCODE_ZENKAKU_HANKAKU";
    /** Key code constant: Japanese alphanumeric key. */
    KeyCode[KeyCode["KEYCODE_EISU"] = 212] = "KEYCODE_EISU";
    /** Key code constant: Japanese non-conversion key. */
    KeyCode[KeyCode["KEYCODE_MUHENKAN"] = 213] = "KEYCODE_MUHENKAN";
    /** Key code constant: Japanese conversion key. */
    KeyCode[KeyCode["KEYCODE_HENKAN"] = 214] = "KEYCODE_HENKAN";
    /** Key code constant: Japanese katakana / hiragana key. */
    KeyCode[KeyCode["KEYCODE_KATAKANA_HIRAGANA"] = 215] = "KEYCODE_KATAKANA_HIRAGANA";
    /** Key code constant: Japanese Yen key. */
    KeyCode[KeyCode["KEYCODE_YEN"] = 216] = "KEYCODE_YEN";
    /** Key code constant: Japanese Ro key. */
    KeyCode[KeyCode["KEYCODE_RO"] = 217] = "KEYCODE_RO";
    /** Key code constant: Japanese kana key. */
    KeyCode[KeyCode["KEYCODE_KANA"] = 218] = "KEYCODE_KANA";
    /** Key code constant: Assist key.
     * Launches the global assist activity.  Not delivered to applications. */
    KeyCode[KeyCode["KEYCODE_ASSIST"] = 219] = "KEYCODE_ASSIST";
    /** Key code constant: Brightness Down key.
     * Adjusts the screen brightness down. */
    KeyCode[KeyCode["KEYCODE_BRIGHTNESS_DOWN"] = 220] = "KEYCODE_BRIGHTNESS_DOWN";
    /** Key code constant: Brightness Up key.
     * Adjusts the screen brightness up. */
    KeyCode[KeyCode["KEYCODE_BRIGHTNESS_UP"] = 221] = "KEYCODE_BRIGHTNESS_UP";
    /** Key code constant: Audio Track key.
     * Switches the audio tracks. */
    KeyCode[KeyCode["KEYCODE_MEDIA_AUDIO_TRACK"] = 222] = "KEYCODE_MEDIA_AUDIO_TRACK";
    /** Key code constant: Sleep key.
     * Puts the device to sleep.  Behaves somewhat like `KEYCODE_POWER` but it
     * has no effect if the device is already asleep. */
    KeyCode[KeyCode["KEYCODE_SLEEP"] = 223] = "KEYCODE_SLEEP";
    /** Key code constant: Wakeup key.
     * Wakes up the device.  Behaves somewhat like `KEYCODE_POWER` but it
     * has no effect if the device is already awake. */
    KeyCode[KeyCode["KEYCODE_WAKEUP"] = 224] = "KEYCODE_WAKEUP";
    /** Key code constant: Pairing key.
     * Initiates peripheral pairing mode. Useful for pairing remote control
     * devices or game controllers, especially if no other input mode is
     * available. */
    KeyCode[KeyCode["KEYCODE_PAIRING"] = 225] = "KEYCODE_PAIRING";
    /** Key code constant: Media Top Menu key.
     * Goes to the top of media menu. */
    KeyCode[KeyCode["KEYCODE_MEDIA_TOP_MENU"] = 226] = "KEYCODE_MEDIA_TOP_MENU";
    /** Key code constant: '11' key. */
    KeyCode[KeyCode["KEYCODE_11"] = 227] = "KEYCODE_11";
    /** Key code constant: '12' key. */
    KeyCode[KeyCode["KEYCODE_12"] = 228] = "KEYCODE_12";
    /** Key code constant: Last Channel key.
     * Goes to the last viewed channel. */
    KeyCode[KeyCode["KEYCODE_LAST_CHANNEL"] = 229] = "KEYCODE_LAST_CHANNEL";
    /** Key code constant: TV data service key.
     * Displays data services like weather, sports. */
    KeyCode[KeyCode["KEYCODE_TV_DATA_SERVICE"] = 230] = "KEYCODE_TV_DATA_SERVICE";
    /** Key code constant: Voice Assist key.
     * Launches the global voice assist activity. Not delivered to applications. */
    KeyCode[KeyCode["KEYCODE_VOICE_ASSIST"] = 231] = "KEYCODE_VOICE_ASSIST";
    /** Key code constant: Radio key.
     * Toggles TV service / Radio service. */
    KeyCode[KeyCode["KEYCODE_TV_RADIO_SERVICE"] = 232] = "KEYCODE_TV_RADIO_SERVICE";
    /** Key code constant: Teletext key.
     * Displays Teletext service. */
    KeyCode[KeyCode["KEYCODE_TV_TELETEXT"] = 233] = "KEYCODE_TV_TELETEXT";
    /** Key code constant: Number entry key.
     * Initiates to enter multi-digit channel nubmber when each digit key is assigned
     * for selecting separate channel. Corresponds to Number Entry Mode (0x1D) of CEC
     * User Control Code. */
    KeyCode[KeyCode["KEYCODE_TV_NUMBER_ENTRY"] = 234] = "KEYCODE_TV_NUMBER_ENTRY";
    /** Key code constant: Analog Terrestrial key.
     * Switches to analog terrestrial broadcast service. */
    KeyCode[KeyCode["KEYCODE_TV_TERRESTRIAL_ANALOG"] = 235] = "KEYCODE_TV_TERRESTRIAL_ANALOG";
    /** Key code constant: Digital Terrestrial key.
     * Switches to digital terrestrial broadcast service. */
    KeyCode[KeyCode["KEYCODE_TV_TERRESTRIAL_DIGITAL"] = 236] = "KEYCODE_TV_TERRESTRIAL_DIGITAL";
    /** Key code constant: Satellite key.
     * Switches to digital satellite broadcast service. */
    KeyCode[KeyCode["KEYCODE_TV_SATELLITE"] = 237] = "KEYCODE_TV_SATELLITE";
    /** Key code constant: BS key.
     * Switches to BS digital satellite broadcasting service available in Japan. */
    KeyCode[KeyCode["KEYCODE_TV_SATELLITE_BS"] = 238] = "KEYCODE_TV_SATELLITE_BS";
    /** Key code constant: CS key.
     * Switches to CS digital satellite broadcasting service available in Japan. */
    KeyCode[KeyCode["KEYCODE_TV_SATELLITE_CS"] = 239] = "KEYCODE_TV_SATELLITE_CS";
    /** Key code constant: BS/CS key.
     * Toggles between BS and CS digital satellite services. */
    KeyCode[KeyCode["KEYCODE_TV_SATELLITE_SERVICE"] = 240] = "KEYCODE_TV_SATELLITE_SERVICE";
    /** Key code constant: Toggle Network key.
     * Toggles selecting broacast services. */
    KeyCode[KeyCode["KEYCODE_TV_NETWORK"] = 241] = "KEYCODE_TV_NETWORK";
    /** Key code constant: Antenna/Cable key.
     * Toggles broadcast input source between antenna and cable. */
    KeyCode[KeyCode["KEYCODE_TV_ANTENNA_CABLE"] = 242] = "KEYCODE_TV_ANTENNA_CABLE";
    /** Key code constant: HDMI #1 key.
     * Switches to HDMI input #1. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_HDMI_1"] = 243] = "KEYCODE_TV_INPUT_HDMI_1";
    /** Key code constant: HDMI #2 key.
     * Switches to HDMI input #2. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_HDMI_2"] = 244] = "KEYCODE_TV_INPUT_HDMI_2";
    /** Key code constant: HDMI #3 key.
     * Switches to HDMI input #3. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_HDMI_3"] = 245] = "KEYCODE_TV_INPUT_HDMI_3";
    /** Key code constant: HDMI #4 key.
     * Switches to HDMI input #4. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_HDMI_4"] = 246] = "KEYCODE_TV_INPUT_HDMI_4";
    /** Key code constant: Composite #1 key.
     * Switches to composite video input #1. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_COMPOSITE_1"] = 247] = "KEYCODE_TV_INPUT_COMPOSITE_1";
    /** Key code constant: Composite #2 key.
     * Switches to composite video input #2. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_COMPOSITE_2"] = 248] = "KEYCODE_TV_INPUT_COMPOSITE_2";
    /** Key code constant: Component #1 key.
     * Switches to component video input #1. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_COMPONENT_1"] = 249] = "KEYCODE_TV_INPUT_COMPONENT_1";
    /** Key code constant: Component #2 key.
     * Switches to component video input #2. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_COMPONENT_2"] = 250] = "KEYCODE_TV_INPUT_COMPONENT_2";
    /** Key code constant: VGA #1 key.
     * Switches to VGA (analog RGB) input #1. */
    KeyCode[KeyCode["KEYCODE_TV_INPUT_VGA_1"] = 251] = "KEYCODE_TV_INPUT_VGA_1";
    /** Key code constant: Audio description key.
     * Toggles audio description off / on. */
    KeyCode[KeyCode["KEYCODE_TV_AUDIO_DESCRIPTION"] = 252] = "KEYCODE_TV_AUDIO_DESCRIPTION";
    /** Key code constant: Audio description mixing volume up key.
     * Louden audio description volume as compared with normal audio volume. */
    KeyCode[KeyCode["KEYCODE_TV_AUDIO_DESCRIPTION_MIX_UP"] = 253] = "KEYCODE_TV_AUDIO_DESCRIPTION_MIX_UP";
    /** Key code constant: Audio description mixing volume down key.
     * Lessen audio description volume as compared with normal audio volume. */
    KeyCode[KeyCode["KEYCODE_TV_AUDIO_DESCRIPTION_MIX_DOWN"] = 254] = "KEYCODE_TV_AUDIO_DESCRIPTION_MIX_DOWN";
    /** Key code constant: Zoom mode key.
     * Changes Zoom mode (Normal, Full, Zoom, Wide-zoom, etc.) */
    KeyCode[KeyCode["KEYCODE_TV_ZOOM_MODE"] = 255] = "KEYCODE_TV_ZOOM_MODE";
    /** Key code constant: Contents menu key.
     * Goes to the title list. Corresponds to Contents Menu (0x0B) of CEC User Control
     * Code */
    KeyCode[KeyCode["KEYCODE_TV_CONTENTS_MENU"] = 256] = "KEYCODE_TV_CONTENTS_MENU";
    /** Key code constant: Media context menu key.
     * Goes to the context menu of media contents. Corresponds to Media Context-sensitive
     * Menu (0x11) of CEC User Control Code. */
    KeyCode[KeyCode["KEYCODE_TV_MEDIA_CONTEXT_MENU"] = 257] = "KEYCODE_TV_MEDIA_CONTEXT_MENU";
    /** Key code constant: Timer programming key.
     * Goes to the timer recording menu. Corresponds to Timer Programming (0x54) of
     * CEC User Control Code. */
    KeyCode[KeyCode["KEYCODE_TV_TIMER_PROGRAMMING"] = 258] = "KEYCODE_TV_TIMER_PROGRAMMING";
    /** Key code constant: Help key. */
    KeyCode[KeyCode["KEYCODE_HELP"] = 259] = "KEYCODE_HELP";
    /** Key code constant: Navigate to previous key.
     * Goes backward by one item in an ordered collection of items. */
    KeyCode[KeyCode["KEYCODE_NAVIGATE_PREVIOUS"] = 260] = "KEYCODE_NAVIGATE_PREVIOUS";
    /** Key code constant: Navigate to next key.
     * Advances to the next item in an ordered collection of items. */
    KeyCode[KeyCode["KEYCODE_NAVIGATE_NEXT"] = 261] = "KEYCODE_NAVIGATE_NEXT";
    /** Key code constant: Navigate in key.
     * Activates the item that currently has focus or expands to the next level of a navigation
     * hierarchy. */
    KeyCode[KeyCode["KEYCODE_NAVIGATE_IN"] = 262] = "KEYCODE_NAVIGATE_IN";
    /** Key code constant: Navigate out key.
     * Backs out one level of a navigation hierarchy or collapses the item that currently has
     * focus. */
    KeyCode[KeyCode["KEYCODE_NAVIGATE_OUT"] = 263] = "KEYCODE_NAVIGATE_OUT";
    /** Key code constant: Primary stem key for Wear
     * Main power/reset button on watch. */
    KeyCode[KeyCode["KEYCODE_STEM_PRIMARY"] = 264] = "KEYCODE_STEM_PRIMARY";
    /** Key code constant: Generic stem key 1 for Wear */
    KeyCode[KeyCode["KEYCODE_STEM_1"] = 265] = "KEYCODE_STEM_1";
    /** Key code constant: Generic stem key 2 for Wear */
    KeyCode[KeyCode["KEYCODE_STEM_2"] = 266] = "KEYCODE_STEM_2";
    /** Key code constant: Generic stem key 3 for Wear */
    KeyCode[KeyCode["KEYCODE_STEM_3"] = 267] = "KEYCODE_STEM_3";
    /** Key code constant: Directional Pad Up-Left */
    KeyCode[KeyCode["KEYCODE_DPAD_UP_LEFT"] = 268] = "KEYCODE_DPAD_UP_LEFT";
    /** Key code constant: Directional Pad Down-Left */
    KeyCode[KeyCode["KEYCODE_DPAD_DOWN_LEFT"] = 269] = "KEYCODE_DPAD_DOWN_LEFT";
    /** Key code constant: Directional Pad Up-Right */
    KeyCode[KeyCode["KEYCODE_DPAD_UP_RIGHT"] = 270] = "KEYCODE_DPAD_UP_RIGHT";
    /** Key code constant: Directional Pad Down-Right */
    KeyCode[KeyCode["KEYCODE_DPAD_DOWN_RIGHT"] = 271] = "KEYCODE_DPAD_DOWN_RIGHT";
    /** Key code constant: Skip forward media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_SKIP_FORWARD"] = 272] = "KEYCODE_MEDIA_SKIP_FORWARD";
    /** Key code constant: Skip backward media key. */
    KeyCode[KeyCode["KEYCODE_MEDIA_SKIP_BACKWARD"] = 273] = "KEYCODE_MEDIA_SKIP_BACKWARD";
    /** Key code constant: Step forward media key.
     * Steps media forward, one frame at a time. */
    KeyCode[KeyCode["KEYCODE_MEDIA_STEP_FORWARD"] = 274] = "KEYCODE_MEDIA_STEP_FORWARD";
    /** Key code constant: Step backward media key.
     * Steps media backward, one frame at a time. */
    KeyCode[KeyCode["KEYCODE_MEDIA_STEP_BACKWARD"] = 275] = "KEYCODE_MEDIA_STEP_BACKWARD";
    /** Key code constant: put device to sleep unless a wakelock is held. */
    KeyCode[KeyCode["KEYCODE_SOFT_SLEEP"] = 276] = "KEYCODE_SOFT_SLEEP";
    /** Key code constant: Cut key. */
    KeyCode[KeyCode["KEYCODE_CUT"] = 277] = "KEYCODE_CUT";
    /** Key code constant: Copy key. */
    KeyCode[KeyCode["KEYCODE_COPY"] = 278] = "KEYCODE_COPY";
    /** Key code constant: Paste key. */
    KeyCode[KeyCode["KEYCODE_PASTE"] = 279] = "KEYCODE_PASTE";
    /** Key code constant: Consumed by the system for navigation up */
    KeyCode[KeyCode["KEYCODE_SYSTEM_NAVIGATION_UP"] = 280] = "KEYCODE_SYSTEM_NAVIGATION_UP";
    /** Key code constant: Consumed by the system for navigation down */
    KeyCode[KeyCode["KEYCODE_SYSTEM_NAVIGATION_DOWN"] = 281] = "KEYCODE_SYSTEM_NAVIGATION_DOWN";
    /** Key code constant: Consumed by the system for navigation left*/
    KeyCode[KeyCode["KEYCODE_SYSTEM_NAVIGATION_LEFT"] = 282] = "KEYCODE_SYSTEM_NAVIGATION_LEFT";
    /** Key code constant: Consumed by the system for navigation right */
    KeyCode[KeyCode["KEYCODE_SYSTEM_NAVIGATION_RIGHT"] = 283] = "KEYCODE_SYSTEM_NAVIGATION_RIGHT";
    /** Key code constant: Show all apps
     */
    KeyCode[KeyCode["KEYCODE_ALL_APPS"] = 284] = "KEYCODE_ALL_APPS";
})(KeyCode || (exports.KeyCode = KeyCode = {}));
