// adb-ts/lib/commands/host-transport/fileSystem/cp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fileSystem_1 = __importDefault(require("../../abstract/fileSystem"));
class CpCommand extends fileSystem_1.default {
    constructor() {
        super(...arguments);
        this.rootCmd = 'cp';
        this.argsMapper = {
            noClobber: '-n',
            symlink: '-s',
            hardLink: '-l',
            noFollowSymlinks: '-P',
            followAllSymlinks: '-L',
            followListedSymlinks: '-H',
            archive: '-a',
            recursive: (_value, options) => options.archive ? '' : '-r',
            noDereference: (_value, options) => options.archive ? '' : '-d',
            preserveTimestamps: (_value, options) => options.archive ? '' : '-p',
            preserve(value) {
                const params = value.all
                    ? ['a']
                    : Object.entries(value)
                        .filter(([key, value]) => key !== 'all' && value)
                        .map(([[char]]) => char);
                return `--preserve=${params.join('')}`;
            },
            delFirst: '-F',
            delDest: '-f',
            update: '-u',
            copyToTarget: '-t'
        };
    }
}
exports.default = CpCommand;
