// adb-ts/lib/commands/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
const parser_1 = require("../parser");
const util_2 = require("../util");
class Command {
    constructor(connection) {
        this.connection = connection;
        this.parser = new parser_1.Parser(this.connection);
    }
    async validateReply(reply) {
        const replyStr = reply.toString();
        switch (replyStr) {
            case util_1.Reply.OKAY:
                return undefined;
            case util_1.Reply.FAIL:
                throw await this.parser.readError();
            default:
                throw this.parser.unexpected(replyStr, [util_1.Reply.OKAY, util_1.Reply.FAIL].join(' or '));
        }
    }
    async readAndValidateReply() {
        return this.validateReply(await this.parser.readAscii(4));
    }
    endConnection() {
        this.connection.end();
    }
    async initAndValidateReply(args) {
        this.connection.write((0, util_2.encodeData)(args));
        try {
            return await this.readAndValidateReply();
        }
        finally {
            this.autoEnd && this.endConnection();
        }
    }
    async initExecute(args) {
        this.connection.write((0, util_2.encodeData)(args));
        try {
            return await this.parser.readAscii(4);
        }
        finally {
            this.autoEnd && this.endConnection();
        }
    }
}
exports.default = Command;
